File: packages/xenium-preprocess/src/xenium_preprocess/r/rctd_prep.R
#!/usr/bin/env Rscript
# ---------------------------------------------------------------------
# rctd_prep.R -- build the RCTD "test object" (Seurat spatial) from the
# mtx/features/barcodes triple + metadata + spatial-coords sidecars
# written by xenium_preprocess.stages.split_prep.
#
# Adapted verbatim from the first half of
#   the internal spatial-RCTD preparation reference
#
# The Rmd's second half (building a spacexr Reference from a scRNA
# mtx triple) is OUT OF SCOPE for step 1 -- the scRNA reference is
# user data, not derived from proseg. Later steps will land that.
#
# Invocation (via xenium_preprocess.stages.rctd_prep):
#   Rscript --vanilla rctd_prep.R \
#       --split-dir=<path> \
#       --stem=<sample_id><name_suffix> \
#       --out-rds=<output.rds> \
#       --assay-name=Proseg \
#       --spatial-key=ST_
#
# Files consumed under <split-dir>:
#   <stem>_counts.mtx.gz
#   <stem>_features.tsv.gz
#   <stem>_barcodes.tsv.gz
#   <stem>_metadata.csv
#   <stem>_spatial_coords.csv.gz
# ---------------------------------------------------------------------

suppressPackageStartupMessages({
  library(Seurat)
  library(Matrix)
  library(readr)
  library(SpatialExperiment)
})

# ----- Arg parsing (mini kv style; no CRAN dep) ----------------------
args <- commandArgs(trailingOnly = TRUE)
kv <- list()
for (a in args) {
  if (grepl("^--[^=]+=", a)) {
    k <- sub("^--", "", sub("=.*$", "", a))
    v <- sub("^[^=]+=", "", a)
    kv[[k]] <- v
  }
}

need <- c("split-dir", "stem", "out-rds", "assay-name", "spatial-key")
missing <- setdiff(need, names(kv))
if (length(missing)) {
  stop("[rctd_prep.R] missing required args: ",
       paste(missing, collapse = ", "))
}

split_dir    <- kv[["split-dir"]]
stem         <- kv[["stem"]]
out_rds      <- kv[["out-rds"]]
assay_name   <- kv[["assay-name"]]
spatial_key  <- kv[["spatial-key"]]

cat(sprintf("[rctd_prep.R] split_dir   = %s\n", split_dir))
cat(sprintf("[rctd_prep.R] stem        = %s\n", stem))
cat(sprintf("[rctd_prep.R] out_rds     = %s\n", out_rds))
cat(sprintf("[rctd_prep.R] assay_name  = %s\n", assay_name))
cat(sprintf("[rctd_prep.R] spatial_key = %s\n", spatial_key))

fp <- function(suffix) file.path(split_dir, paste0(stem, suffix))
meta_path    <- fp("_metadata.csv")
coords_path  <- fp("_spatial_coords.csv.gz")
mtx_path     <- fp("_counts.mtx.gz")
feats_path   <- fp("_features.tsv.gz")
barcodes_path <- fp("_barcodes.tsv.gz")

for (p in c(meta_path, coords_path, mtx_path, feats_path, barcodes_path)) {
  if (!file.exists(p)) {
    stop("[rctd_prep.R] missing input file: ", p)
  }
}

# ----- Load the SPLIT triple + sidecars (mirrors the Rmd) ------------
meta   <- read.csv(meta_path, row.names = 1, check.names = FALSE)
coords <- as.matrix(readr::read_csv(coords_path, show_col_types = FALSE))

mtx <- ReadMtx(
  mtx           = mtx_path,
  features      = feats_path,
  cells         = barcodes_path,
  feature.column = 1,   # single-column features file (see split_prep.py)
  cell.column    = 1
)

# ----- Build the SpatialExperiment (matches the Rmd) -----------------
spe <- SpatialExperiment(
  assays  = list(counts = mtx),
  colData = DataFrame(meta[colnames(mtx), , drop = FALSE])
)

spatialCoords(spe) <- as.matrix(coords)

# Keep only the (x, y) columns for the spatial coords. The reference
# Rmd assumes the coords CSV has an id column + x + y; readr::read_csv
# reads the id into the first column, so we hard-select "x" and "y".
if (all(c("x", "y") %in% colnames(spatialCoords(spe)))) {
  spatialCoords(spe) <- spatialCoords(spe)[, c("x", "y"), drop = FALSE]
}

# ----- Convert to Seurat + attach spatial DimReduc (matches the Rmd) -
seu <- CreateSeuratObject(
  counts    = counts(spe),
  assay     = assay_name,
  meta.data = as.data.frame(colData(spe))
)

sp_coords <- spatialCoords(spe)
# Match the Rmd's rename to ST_1 / ST_2 for the DimReduc.
colnames(sp_coords) <- c(paste0(spatial_key, 1), paste0(spatial_key, 2))
rownames(sp_coords) <- colnames(seu)

seu[["spatial"]] <- CreateDimReducObject(
  sp_coords, assay = assay_name, key = spatial_key
)

seu$x <- sp_coords[, 1]
seu$y <- sp_coords[, 2]

# ----- Save (drop the SPE to keep the RDS small) --------------------
rm(spe)
gc()

dir.create(dirname(out_rds), showWarnings = FALSE, recursive = TRUE)
saveRDS(seu, file = out_rds)

cat(sprintf("[rctd_prep.R] wrote %s (cells=%d, features=%d)\n",
            out_rds, ncol(seu), nrow(seu)))

File: packages/xenium-preprocess/src/xenium_preprocess/stages/rctd_prep.py
"""Stage 4: SPLIT triple → RCTD test object (RDS).

Adapted from
    the internal spatial-RCTD preparation reference

The reference Rmd reads the mtx/features/barcodes triple + the two
sidecars written by stage 3, builds a SpatialExperiment, converts it
to a Seurat object with a "Proseg" assay + a spatial DimReduc keyed
"ST_", and saves as an RDS. That RDS is the RCTD "test object" —
downstream `create.RCTD(spatial_seurat, reference)` calls take this
as the query side of the deconvolution.

The Rmd's second half (building a spacexr `Reference` from a scRNA
mtx triple) is OUT OF SCOPE for xenium-preprocess — the scRNA reference is user
data, not derived from proseg. That part will land in a later step.

Which anndata layer feeds RCTD: `rctd_prep.source_layer` in the config
(default `maxpost_counts` — user request 2026-07-10, (internal issue review).
The choice is applied UPSTREAM at split_prep (pipeline.py threads
`rctd_prep.source_layer` into `run_split_prep(layer=...)`), so this
module itself does not touch the h5ad — it just consumes the mtx
already exported by split_prep. RCTD's `create.RCTD(...,
require_int=TRUE)` requires integer counts, which is why maxpost is
the default (the alternative `expected_counts` layer is continuous).

This module is a thin Python wrapper: it locates the R script that
ships alongside the package (`xenium_preprocess/r/rctd_prep.R`),
shells out to `Rscript`, threads through the resolved paths + args,
and fails loud on non-zero exit.

The R side needs Seurat + Matrix + readr + SpatialExperiment (no
spacexr/SPLIT — those are only needed by ref-build/rctd-split). Pass
`--r-lib-paths` (config key `r_lib_paths`) pointing at a renv/user
library if any of those aren't on the default R library search path;
this stage prepends those paths to the child Rscript's `R_LIBS_USER`
(mirrors rctd-split's `r_lib_paths` mechanism — see
`rctd_split/stages/export_mtx.py`).
"""
from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path

from xenium_preprocess._internal.compat import sentinel_exists
from xenium_preprocess._internal.layout import rctd_path
from xenium_preprocess._internal.logging import log


def _package_r_script() -> Path:
    """Absolute path to the R script shipped in the package."""
    return Path(__file__).resolve().parent.parent / "r" / "rctd_prep.R"


def run_rctd_prep(
    sample_id: str,
    run_id: str,
    split_dir: Path,
    output_root: Path,
    name_suffix: str,
    rscript_bin: str,
    assay_name: str,
    spatial_key: str,
    force_rerun: bool,
    r_lib_paths: list[str] | list[Path] | None = None,
) -> Path:
    """Invoke `Rscript rctd_prep.R` on the SPLIT triple.

    Returns the path to the output RDS.
    """
    # Final locked layout: rctd/<S>_test_object.rds under the run folder.
    out_rds = rctd_path(output_root, sample_id, run_id, "test_object")
    out_rds.parent.mkdir(parents=True, exist_ok=True)

    if sentinel_exists(out_rds, force_rerun):
        log(f"[rctd_prep] sentinel exists: {out_rds} — skipping "
            f"(pass --force-rerun to re-run).")
        return out_rds

    # Locate Rscript. shutil.which fails loud vs. a silent NotFound
    # deep in subprocess.
    if shutil.which(rscript_bin) is None:
        raise SystemExit(
            f"[rctd_prep] Rscript binary not found on PATH: {rscript_bin!r}. "
            "Options: (a) `module load fhR/4.4.1-foss-2023b` before invoking "
            "the pipeline; (b) pass --rscript-bin /absolute/path/to/Rscript; "
            "(c) set config.rctd_prep.rscript_bin to the explicit binary path."
        )

    r_script = _package_r_script()
    if not r_script.exists():
        raise SystemExit(
            f"[rctd_prep] R script missing from the package: {r_script}. "
            "Something's off with the install; expected r/rctd_prep.R "
            "under the xenium_preprocess package tree."
        )

    stem = f"{sample_id}{name_suffix}"
    args = [
        rscript_bin, "--vanilla",
        str(r_script),
        f"--split-dir={split_dir}",
        f"--stem={stem}",
        f"--out-rds={out_rds}",
        f"--assay-name={assay_name}",
        f"--spatial-key={spatial_key}",
    ]
    # R silently SKIPS an R_LIBS_USER entry that doesn't exist instead
    # of erroring — validate up front so a typo'd/stale --r-lib-paths
    # fails loud here rather than surfacing as a confusing
    # "there is no package called ..." deep inside the R script.
    env = os.environ.copy()
    if r_lib_paths:
        missing = [str(p) for p in r_lib_paths if not Path(p).is_dir()]
        if missing:
            raise SystemExit(
                f"[rctd_prep] --r-lib-paths / r_lib_paths entry does not "
                f"exist: {missing}. R silently ignores a missing "
                "R_LIBS_USER directory, so this fails loud instead."
            )
        prepend = ":".join(str(p) for p in r_lib_paths)
        existing = env.get("R_LIBS_USER", "")
        env["R_LIBS_USER"] = f"{prepend}:{existing}" if existing else prepend
        log(f"[rctd_prep] R_LIBS_USER prepended with: {prepend}")

    log(f"[rctd_prep] launching: {' '.join(args)}")
    # Stream both stdout + stderr into the pipeline log (subprocess
    # inherits our stdout/stderr fds by default). check=True raises on
    # non-zero exit; the resulting CalledProcessError carries the code.
    subprocess.run(args, check=True, env=env)

    if not out_rds.exists():
        raise SystemExit(
            f"[rctd_prep] Rscript exited 0 but the expected output "
            f"was not written: {out_rds}."
        )
    log(f"[rctd_prep] wrote {out_rds}")
    return out_rds
